File: src/main/java/com/company/erp/features/chats/controller/MessageController.java
package com.company.erp.features.chats.controller;

import com.company.erp.core.database.PageQuery;
import com.company.erp.core.response.PageResponse;
import com.company.erp.core.security.AuthenticatedUser;
import com.company.erp.core.security.Permissions;
import com.company.erp.features.chats.dto.*;
import com.company.erp.features.chats.entity.Message;
import com.company.erp.features.chats.entity.MessageReaction;
import com.company.erp.features.chats.service.ConversationService;
import com.company.erp.features.chats.service.MessageService;
import com.company.erp.features.chats.ws.ChatBroadcaster;
import jakarta.validation.Valid;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1/chats")
@PreAuthorize("hasAuthority('" + Permissions.CHAT_READ + "')")
public class MessageController {

    private final MessageService messages;
    private final ConversationService conversations;
    private final ChatBroadcaster broadcaster;

    public MessageController(MessageService messages,
                             ConversationService conversations,
                             ChatBroadcaster broadcaster) {
        this.messages = messages;
        this.conversations = conversations;
        this.broadcaster = broadcaster;
    }

    @GetMapping("/conversations/{convId}/messages")
    public PageResponse<MessageDto> history(
            @PathVariable Long convId,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "30") int pageSize,
            @RequestParam(required = false) String sort) {
        Long me = AuthenticatedUser.require().userId();
        return PageResponse.from(
                messages.history(convId, me, new PageQuery(page, pageSize, null, sort)),
                m -> MessageDto.from(m, reactionDtos(m.getId())));
    }

    @GetMapping("/conversations/{convId}/messages/search")
    public PageResponse<MessageDto> search(
            @PathVariable Long convId,
            @RequestParam String q,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "30") int pageSize) {
        Long me = AuthenticatedUser.require().userId();
        return PageResponse.from(
                messages.search(convId, me, q, new PageQuery(page, pageSize, q, null)),
                m -> MessageDto.from(m, reactionDtos(m.getId())));
    }

    @PostMapping("/conversations/{convId}/messages")
    @PreAuthorize("hasAuthority('" + Permissions.CHAT_WRITE + "')")
    public MessageDto send(@PathVariable Long convId, @Valid @RequestBody SendMessageRequest body) {
        Long me = AuthenticatedUser.require().userId();
        Message m = messages.send(convId, me, body);
        MessageDto dto = MessageDto.from(m, List.of());
        broadcaster.toConversation(convId, "message.send", dto);
        broadcaster.toUsers(conversations.memberUserIds(convId), "inbox", "message.send", dto);
        return dto;
    }

    @PatchMapping("/messages/{id}")
    @PreAuthorize("hasAuthority('" + Permissions.CHAT_WRITE + "')")
    public MessageDto edit(@PathVariable Long id, @Valid @RequestBody EditMessageRequest body) {
        Long me = AuthenticatedUser.require().userId();
        Message m = messages.edit(id, me, body);
        MessageDto dto = MessageDto.from(m, reactionDtos(id));
        broadcaster.toConversation(m.getConversationId(), "message.edit", dto);
        return dto;
    }

    @DeleteMapping("/messages/{id}")
    @PreAuthorize("hasAuthority('" + Permissions.CHAT_WRITE + "')")
    public MessageDto delete(@PathVariable Long id) {
        Long me = AuthenticatedUser.require().userId();
        Message m = messages.delete(id, me);
        MessageDto dto = MessageDto.from(m, List.of());
        broadcaster.toConversation(m.getConversationId(), "message.delete", dto);
        return dto;
    }

    @PostMapping("/messages/{id}/reactions")
    @PreAuthorize("hasAuthority('" + Permissions.CHAT_WRITE + "')")
    public List<ReactionDto> toggleReaction(@PathVariable Long id,
                                            @Valid @RequestBody ToggleReactionRequest body) {
        Long me = AuthenticatedUser.require().userId();
        List<MessageReaction> updated = messages.toggleReaction(id, me, body.emoji());
        List<ReactionDto> dtos = updated.stream().map(ReactionDto::from).toList();
        Message m = messages.getById(id);
        broadcaster.toConversation(m.getConversationId(), "reaction.toggle",
                java.util.Map.of("messageId", id, "reactions", dtos));
        return dtos;
    }

    private List<ReactionDto> reactionDtos(Long messageId) {
        return messages.reactionsFor(messageId).stream().map(ReactionDto::from).toList();
    }
}

File: app/Features/Chat/Controllers/MessageController.php
<?php

namespace App\Features\Chat\Controllers;

use App\Features\Chat\Dto\MessageDto;
use App\Features\Chat\Dto\ReactionDto;
use App\Features\Chat\Models\ChatMessage;
use App\Features\Chat\Models\ConversationMember;
use App\Features\Chat\Requests\EditMessageRequest;
use App\Features\Chat\Requests\SendMessageRequest;
use App\Features\Chat\Requests\ToggleReactionRequest;
use App\Features\Chat\Services\ChatBroadcaster;
use App\Features\Chat\Services\ConversationService;
use App\Features\Chat\Services\MessageService;
use App\Http\Controllers\Controller;
use App\Support\Exceptions\BadRequestException;
use App\Support\Pagination\PageQuery;
use App\Support\Pagination\PageResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Auth;

class MessageController extends Controller
{
    /** @var MessageService */
    private $messages;

    /** @var ConversationService */
    private $conversations;

    /** @var ChatBroadcaster */
    private $broadcaster;

    public function __construct(
        MessageService $messages,
        ConversationService $conversations,
        ChatBroadcaster $broadcaster
    ) {
        $this->messages = $messages;
        $this->conversations = $conversations;
        $this->broadcaster = $broadcaster;
    }

    /** Paginated message history for a conversation, newest-first, with read receipts. */
    public function history(int $convId, Request $request)
    {
        $me = (int) Auth::guard('api')->id();
        $memberLastRead = $this->memberLastReadMap($convId);
        $paged = $this->messages->history($convId, $me, PageQuery::fromRequest($request));

        $self = $this;
        return PageResponse::from($paged, function (ChatMessage $m) use ($memberLastRead, $self) {
            return MessageDto::from($m, $self->reactionDtos((int) $m->id), $self->readByForMessage($m, $memberLastRead));
        });
    }

    /** Case-insensitive substring search on message body within a conversation. */
    public function search(int $convId, Request $request)
    {
        $me = (int) Auth::guard('api')->id();
        $q = $request->query('q');
        if ($q === null) {
            throw new BadRequestException('Query parameter q is required');
        }
        $memberLastRead = $this->memberLastReadMap($convId);
        $query = new PageQuery((int) $request->query('page', 1), (int) $request->query('pageSize', 30), $q, null);
        $paged = $this->messages->search($convId, $me, $q, $query);

        $self = $this;
        return PageResponse::from($paged, function (ChatMessage $m) use ($memberLastRead, $self) {
            return MessageDto::from($m, $self->reactionDtos((int) $m->id), $self->readByForMessage($m, $memberLastRead));
        });
    }

    /** Send a text / image / voice / file message; broadcasts to the conv topic + inbox. */
    public function send(int $convId, SendMessageRequest $request)
    {
        $me = (int) Auth::guard('api')->id();
        $m = $this->messages->send($convId, $me, $request->validated());
        $dto = MessageDto::from($m, []);
        $this->broadcaster->toConversation($convId, 'message.send', $dto);
        $this->broadcaster->toUsers($this->conversations->memberUserIds($convId), 'inbox', 'message.send', $dto);
        return $dto;
    }

    /** Edit own TEXT message within 15 minutes of sending. */
    public function edit(int $id, EditMessageRequest $request)
    {
        $me = (int) Auth::guard('api')->id();
        $m = $this->messages->edit($id, $me, $request->validated());
        $dto = MessageDto::from($m, $this->reactionDtos($id));
        $this->broadcaster->toConversation((int) $m->conversation_id, 'message.edit', $dto);
        return $dto;
    }

    /** Soft-delete a message (sender only); body & attachment fields nulled in responses. */
    public function delete(int $id)
    {
        $me = (int) Auth::guard('api')->id();
        $m = $this->messages->delete($id, $me);
        $dto = MessageDto::from($m, []);
        $this->broadcaster->toConversation((int) $m->conversation_id, 'message.delete', $dto);
        return $dto;
    }

    /** Toggle an emoji reaction on a message; returns the full reactions list afterwards. */
    public function toggleReaction(int $id, ToggleReactionRequest $request)
    {
        $me = (int) Auth::guard('api')->id();
        $data = $request->validated();
        $updated = $this->messages->toggleReaction($id, $me, $data['emoji']);
        $dtos = $updated->map([ReactionDto::class, 'from'])->values()->all();
        $m = $this->messages->getById($id);
        $this->broadcaster->toConversation((int) $m->conversation_id, 'reaction.toggle', [
            'messageId' => $id,
            'reactions' => $dtos,
        ]);
        return $dtos;
    }

    /**
     * @return array<int, array<string, mixed>>
     */
    public function reactionDtos(int $messageId): array
    {
        return $this->messages->reactionsFor($messageId)->map([ReactionDto::class, 'from'])->values()->all();
    }

    /**
     * userId → that member's last-read message id (0 if never read anything).
     *
     * @return array<int, int>
     */
    private function memberLastReadMap(int $convId): array
    {
        $map = [];
        foreach (ConversationMember::query()->where('conversation_id', $convId)->get() as $m) {
            $map[(int) $m->user_id] = $m->last_read_message_id === null ? 0 : (int) $m->last_read_message_id;
        }
        return $map;
    }

    /**
     * All members (excluding sender) whose lastReadMessageId >= this message's id.
     *
     * @param array<int, int> $memberLastRead
     * @return array<int, int>
     */
    public function readByForMessage(ChatMessage $m, array $memberLastRead): array
    {
        $out = [];
        foreach ($memberLastRead as $uid => $lastRead) {
            if ((int) $uid === (int) $m->sender_id) {
                continue;
            }
            if ($lastRead >= (int) $m->id) {
                $out[] = (int) $uid;
            }
        }
        return $out;
    }
}
